pkgs/servers/web-apps/croodle/default.nix

```diff
@@ -0,0 +1,67 @@
+{ lib
+, stdenvNoCC
+, writeText
+, nixosTests
+, dataDir ? "/var/lib/croodle"
+, debug ? false
+}:
+
+let
+  configFile = writeText "croodle-config.php" ''
+    <?php
+
+    return array(
+      /*
+       * dataDir (String)
+       * relative or absolute path to folder where polls are stored
+       */
+      'dataDir' => getenv('CROODLE__DATA_DIR') ?: '${dataDir}/',
+
+      /*
+       * debug (Boolean)
+       * controls Slim debug mode
+       */
+      'debug' => getenv('CROODLE__DEBUG') ?: ${lib.boolToString debug}
+    );
+  '';
+
+in
+
+stdenvNoCC.mkDerivation rec {
```

Comment:
I think we should use the `finalAttrs` pattern here.

Let me share a couple of useful links regarding this:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Reply:
Do you think the derivation arguments should be moved into the attribute set as well?

Reply:
Can you please elaborate?

Reply:
Sure thing. Right now, the `dataDir` and `debug` options can be overridden because they appear in the derivation argument set:  
```
{ lib
, fetchurl
, nixosTests
, stdenvNoCC
, writeText
, dataDir ? "/var/lib/croodle"
, debug ? false
}:

let
  configFile = ...
in
stdenvNoCC.mkDerivation (finalAttrs: { ...
```

If I'm understanding the finalAttrs pattern, I could also do the following:

```
{ lib
, fetchurl
, nixosTests
, stdenvNoCC
}:

stdenvNoCC.mkDerivation (finalAttrs: { ...
  dataDir = "/var/lib/croodle";
  debug = false;
```

Then I could access the values of `finalAttrs.dataDir` and `finalAttrs.debug` during the patch phase to write the configuration file. They would still be overrideable because of the derivation wrapped in a lambda works.

Is my understanding correct and would you suggest I do so?